### PR TITLE
Pivotal ID # 173569202: Empty Section Parent Cell

### PR DESCRIPTION
**Pull request by jhoanmanuelms**
https://www.pivotaltracker.com/story/show/173569202

Improve excel reader:
- When all the cells in a row are empty, this will be interpreted as an empty line
- Remove empty elements from the end of a row
- Remove all empty rows starting at the last populated
- Add unit tests for these edge cases

Fixes # 173569202

### CODE REVIEW

File: commons/commons-util/src/main/kotlin/ebi/ac/uk/dsl/excel/ExcelSheet.kt [hunk 1]
```diff
@@ -1,13 +1,17 @@
...
+import ebi.ac.uk.base.EMPTY
 import org.apache.poi.xssf.streaming.SXSSFSheet
 
 class ExcelSheet(private val sheet: SXSSFSheet) {
     private val rowIterator = buildIterator()
 
     fun row(cellsBuilder: ExcelRow.() -> Unit) = ExcelRow(newRow()).apply { cellsBuilder() }
 
-    fun emptyRow() = row { }
+    fun emptyRow() = row {
```
Comment (Juan-EBI) [inline commons/commons-util/src/main/kotlin/ebi/ac/uk/dsl/excel/ExcelSheet.kt:11]: this is only used as part of test/itest , IMO it should be in  test class package or function

File: commons/commons-util/src/main/kotlin/ebi/ac/uk/util/file/PoiExt.kt [hunk 2]
```diff
@@ -5,20 +5,19 @@ import org.apache.poi.ss.usermodel.Cell
...
 import org.apache.poi.ss.usermodel.Sheet
-import java.util.LinkedList
 
 /**
  * Retrieves the sheet representation as a list of tsv separated string List. Note that as stream reader ignored empty
  * rows they are completed with emtpy values.
  */
 fun Sheet.asTsvList(): List<String> {
-    val elements = LinkedList<String>()
+    val elements = mutableListOf<String>()
```
Comment (Juan-EBI) [inline commons/commons-util/src/main/kotlin/ebi/ac/uk/util/file/PoiExt.kt:14]: mutable list in kotlin is by default a ArrayList which performance for remove operations is terrible. (as backing array need to be re created)

File: commons/commons-util/src/main/kotlin/ebi/ac/uk/util/file/PoiExt.kt [hunk 3]
```diff
@@ -27,7 +26,21 @@ private fun Row.asString(): String {
...
+    }
+}
+
+private fun MutableList<String>.trim(): MutableList<String> {
+    if (last().isBlank()) {
+        for (idx in indices.reversed()) {
+            if (elementAt(idx).isNotBlank()) break
+            removeAt(idx)
+        }
+    }
```
Comment (Juan-EBI) [inline commons/commons-util/src/main/kotlin/ebi/ac/uk/util/file/PoiExt.kt:41]: this for looks overcomplex, definitely kotlin have a cleaner declarative approach. What about

```
this.asReversed().dropWhile { it.isBlank() }
```

File: commons/commons-util/src/test/kotlin/ebi/ac/uk/util/file/ExcelReaderTest.kt [hunk 4]
```diff
@@ -30,6 +30,8 @@ class ExcelReaderTest(private val temporaryFolder: TemporaryFolder) {
                 row {
                     cell("Study")
                     cell("SECT-001")
+                    cell("")
```
Comment (Juan-EBI) [inline commons/commons-util/src/test/kotlin/ebi/ac/uk/util/file/ExcelReaderTest.kt:33]: here we are complicating main test case to include corner case scenarios testing, I think instead, we should add an extra simple test case `when extra empty cells` and keep this one only for happy path.

File: commons/commons-util/src/main/kotlin/ebi/ac/uk/dsl/excel/ExcelSheet.kt (hunk 1 above)
Comment (jhoanmanuelms) [reply commons/commons-util/src/main/kotlin/ebi/ac/uk/dsl/excel/ExcelSheet.kt:11]: All DSLs we have are used for testing, perhaps we could create a separate module in another PR and move them all? Right now they're in `commons-util` but we could have something like `commons-dsl` ?

File: commons/commons-util/src/main/kotlin/ebi/ac/uk/util/file/PoiExt.kt (hunk 2 above)
Comment (jhoanmanuelms) [reply commons/commons-util/src/main/kotlin/ebi/ac/uk/util/file/PoiExt.kt:14]: Reverted

File: commons/commons-util/src/main/kotlin/ebi/ac/uk/util/file/PoiExt.kt (hunk 3 above)
Comment (jhoanmanuelms) [reply commons/commons-util/src/main/kotlin/ebi/ac/uk/util/file/PoiExt.kt:41]: Thanks! I knew there was a better way. You're like my Kotlin linter :laughing: 

Only one detail, I ended up using: `dropLastWhile { it.isBlank() }`

File: commons/commons-util/src/test/kotlin/ebi/ac/uk/util/file/ExcelReaderTest.kt (hunk 4 above)
Comment (jhoanmanuelms) [reply commons/commons-util/src/test/kotlin/ebi/ac/uk/util/file/ExcelReaderTest.kt:33]: Done
